roachtest: add an Operation for network failures

andrewbaptist opened the pull request:
We want to be able to test network partitions, but full and partial as
part of our DRT cluster. This commit adds a failure mode for network
partitions. Half the time it does a partial partition and the other half
it does a full partition.

Fixes: https://github.com/cockroachdb/cockroach/issues/121940

Epic: none

Release note: None

Commit:
roachtest: add an Operation for network failures

We want to be able to test network partitions, but full and partial as
part of our DRT cluster. This commit adds a failure mode for network
partitions. Half the time it does a partial partition and the other half
it does a full partition.

Fixes: #121940

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/122157)
<!-- Reviewable:end -->

andrewbaptist commented:
@itsbilal can you help me figure out how to test this Operation to make sure it works? I'm not sure if I can use this on a standard roachprod cluster.

Commit:
roachtest: fix usage of c.t

The cluster test is not set up in all cases. Specifically in the case of
the new operators framework. Ideally we should set this up to avoid
losing this logging so adding a TODO as well.

Epic: none

Release note: None

andrewbaptist commented:
@itsbilal I've validated that this correctly creates partial and full partitions on cluster and that it cleans them up afterwards. I realize the `Dependencies` piece isn't yet implemented, and I'm not sure how this all runs. Let me know if you want to chat about getting this rolled out to DRT or if it just happens automatically once this is merged.

Commit:
roachtest: run Cleanup after Ctrl-C

Previously if `roachtest run-operation ...` was run and the user
terminated the process with a Ctrl-C, the Cleanup of the Operation was
not run. In the case of network partitions, this would leave the
partition in place. This change removes the early exit when Ctrl-C is
pressed by falling through and calling Cleanup with a fresh Context.
This is more expected behavior.

Epic: none

Release note: None

itsbilal commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -2365,7 +2365,7 @@ func (c *clusterImpl) RunE(ctx context.Context, options install.RunOptions, args
 
 	cmd := strings.Join(args, " ")
-	c.t.L().Printf("running cmd `%s` on nodes [%v]; details in %s.log", roachprod.TruncateString(cmd, 30), nodes, logFile)
+	l.Printf("running cmd `%s` on nodes [%v]; details in %s.log", roachprod.TruncateString(cmd, 30), nodes, logFile)
```
I think this is slightly inaccurate as we want the main logger for the test/operation, not the specific separate log file for the command. For now I think we should just gate this behind `if c.t != nil {...}`.

itsbilal commented on pkg/cmd/roachtest/operations/network_partition.go:
```diff
@@ -0,0 +1,114 @@
+) registry.OperationCleanup {
+	nodeCount := c.Spec().NodeCount
+	if nodeCount <= 1 {
+		o.Status("not enough nodes to create a partition")
```
same here, this can be an o.Fatal

itsbilal commented on pkg/cmd/roachtest/operations/network_partition.go:
```diff
@@ -0,0 +1,114 @@
+// partition operations.
+func registerNetworkPartition(r registry.Registry) {
+	r.AddOperation(registry.OperationSpec{
+		Name:             "network-full-partition",
```
nit: `network-partition/full` and `network-partition/partial` ?

itsbilal commented on pkg/cmd/roachtest/operations/network_partition.go:
```diff
@@ -0,0 +1,114 @@
+		Owner:            registry.OwnerKV,
+		Timeout:          1 * time.Minute,
+		CompatibleClouds: registry.AllClouds,
+		Dependencies:     []registry.OperationDependency{registry.OperationRequiresZeroUnavailableRanges},
```
as discussed, this can be `OperationRequiresZeroUnderreplicatedRanges`.

itsbilal commented on pkg/cmd/roachtest/run.go:
```diff
@@ -517,12 +517,11 @@ func runOperation(register func(registry.Registry), filter string, clusterName s
 	op.Status(fmt.Sprintf("operation ran successfully; waiting %s before cleanup", roachtestflags.WaitBeforeCleanup))
 	select {
 	case <-ctx.Done():
-		return ctx.Err()
```
thanks for doing this! might be worth adding a comment here that this is to get Ctrl+C to run cleanup.

itsbilal reviewed (commented):
Thanks for doing this! This is looking great, just some minor comments

andrewbaptist reviewed (commented):
TFTR!

BTW - it would be nice if there was a way to "list" available operators using `run-operation`. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122157#-:-NvhLqck2yplX5yxevrv:b-7z87ka)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @itsbilal, and @srosenberg)

___
*[`pkg/cmd/roachtest/cluster.go` line 2368 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/122157#gh-1569286732:-NvhLuME2RBrJR7Q9fcc:bja7j4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c518e7c60bd8f91cfa76d15a95cba2751b67e996/pkg/cmd/roachtest/cluster.go#L2368)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I think this is slightly inaccurate as we want the main logger for the test/operation, not the specific separate log file for the command. For now I think we should just gate this behind `if c.t != nil {...}`.
</blockquote></details>

Done (added a TODO for you as well :)
___
*[`pkg/cmd/roachtest/run.go` line 0 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/122157#gh-1569289680:-NvhLwsJDgEdb3htGW9T:b-sbf2ja) ([raw file](https://github.com/cockroachdb/cockroach/blob/c518e7c60bd8f91cfa76d15a95cba2751b67e996/pkg/cmd/roachtest/run.go#L0)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/cockroachdb/cockroach/pull/122157#discussion_r1569289680) to the right spot — sorry!)*

thanks for doing this! might be worth adding a comment here that this is to get Ctrl+C to run cleanup.
</blockquote></details>

I added a comment.
___
*[`pkg/cmd/roachtest/operations/network_partition.go` line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/122157#gh-1569287622:-NvhQ-9SETOPvv3qX6e5:b-sbuihr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c518e7c60bd8f91cfa76d15a95cba2751b67e996/pkg/cmd/roachtest/operations/network_partition.go#L43)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

this can be an o.Fatal
</blockquote></details>

Changed. 
___
*[`pkg/cmd/roachtest/operations/network_partition.go` line 79 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/122157#gh-1569287878:-NvhQ8AG7UlNJVvFlsSM:b499q7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c518e7c60bd8f91cfa76d15a95cba2751b67e996/pkg/cmd/roachtest/operations/network_partition.go#L79)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

same here, this can be an o.Fatal
</blockquote></details>

Changed
___
*[`pkg/cmd/roachtest/operations/network_partition.go` line 99 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/122157#gh-1569288710:-NvhQY3R6Jtgl6taEvW0:b499q7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c518e7c60bd8f91cfa76d15a95cba2751b67e996/pkg/cmd/roachtest/operations/network_partition.go#L99)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

nit: `network-partition/full` and `network-partition/partial` ?
</blockquote></details>

Changed
___
*[`pkg/cmd/roachtest/operations/network_partition.go` line 103 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/122157#gh-1569289106:-NvhQXO8CJVScHJR0tq4:b499q7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c518e7c60bd8f91cfa76d15a95cba2751b67e996/pkg/cmd/roachtest/operations/network_partition.go#L103)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

as discussed, this can be `OperationRequiresZeroUnderreplicatedRanges`.
</blockquote></details>

Changed


<!-- Sent from Reviewable.io -->

itsbilal reviewed (approved):
LGTM, thanks a ton!

andrewbaptist commented:
bors r=itsbilal

TFTR! I'm excited to see this deployed. Let me know if there is anything that needs to happen next (and if this needs to be backported to 24.1)
